src/ec/curve25519/x25519.rs

```diff
@@ -158,8 +158,10 @@ fn x25519_ecdh(
     Ok(())
 }
 
-#[cfg(all(not(target_os = "ios"), target_arch = "arm"))]
+#[cfg(ring_x25519_neon)]
 fn x25519_neon(out: &mut ops::EncodedPoint, scalar: &ops::MaskedScalar, point: &ops::EncodedPoint) {
+    const _NOT_APPLE_COMPATIBLE: () = assert!(!cfg!(target_vendor = "apple"));
```

Comment:
This doesn't make sense out of the gate, at least to me. Was the idea to followup and allow some Apple targets to use this? Even with that dropping M{1, 2, 3} in the middle comes off weird and puts main in a odd state (see below about the feasibility of `target_abi` gating). 

However I noticed this does fix a possible bug where 32-bit watchOS may try and use this incorrectly.

> This is a temporary measure until we can properly condition it on the target ABI.

I don't really think the ABI is needed, target arch and OS get the whole way to what you seem to want? And then `#[cfg(target_abi = "")]` [isn't even stable yet](https://github.com/rust-lang/rust/pull/119590) and probably won't be for a while so relying on it for a new release is foggy. This I'm less sure of, since I don't have context, but this plan might also stumble over the simulator targets that have `target_abi="sim"` without a workaround.

The only Apple device's that don't support NEON are 32-bit iOS (`armv7s)` and actually-32-bit watchOS (`armv7k`). 64-bit-but-half-pointers (`arm64_32`) and the other 64-bit watchOS target (`aarch64`) have it available. Those are the two targets that need excluded vs many safe-to-assume combos for Apple's platforms.

With the specifics covered, does _ring_'s policy forbid turning this on for any target with `cfg(target_feature = "neon")`? It seems like the simplest approach (if you're not worried about CPU bugs) and the assembly can be tested on another aarch64 device for correctness (register restoring, etc).

Reply:
The problem is that x25519_NEON's assembly language code doesn't work in Apple's OS's because it isn't ABI-compatible with them, I think because it uses a register that the OS reserves.

This only affects 32-bit targets (target_arch = "arm"), not 64-bit or arm64-32.

Reply:
> The problem is that x25519_NEON's assembly language code doesn't work in Apple's OS's because it isn't ABI-compatible with them, I think because it uses a register that the OS reserves.
> 
> This only affects 32-bit targets (target_arch = "arm"), not 64-bit or arm64-32.

Not sure about conventions, but this seems like potentially useful insight to sum up in a comment in the code.

Reply:
I added a comment calling out the ABI-incompatibility, but without the speculation about the register usage, as I didn't find a definitive reason.

Reply:
RE: target_abi not being stable, and target_abi = sim, those are all good points. target_abi not being stable is one of the main reasons we have to condition on the target vendor.